Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp

```diff
@@ -3147,6 +3246,244 @@ auto B3IRGenerator::addCallRef(const TypeDefinition& signature, Vector<Expressio
+                // We pessimistically assume we could be calling to something that is bounds checking.
+                // FIXME: We shouldn't have to do this: https://bugs.webkit.org/show_bug.cgi?id=172181
+                patchpoint->clobberLate(PinnedRegisterInfo::get().toSave(MemoryMode::BoundsChecking));
+                patchpoint->setGenerator([this, handle, returnType](CCallHelpers& jit, const B3::StackmapGenerationParams& params) {
+                    AllowMacroScratchRegisterUsage allowScratch(jit);
+                    handle->generate(jit, params, this);
+                    jit.call(params[params.proc().resultCount(returnType)].gpr(), WasmEntryPtrTag);
+                });
+            }));
+        UpsilonValue* embedderCallResultUpsilon = returnType == B3::Void ? nullptr : isEmbedderBlock->appendNew<UpsilonValue>(m_proc, origin(), embedderCallResult);
+        isEmbedderBlock->appendNewControlValue(m_proc, Jump, origin(), continuation);
+
+        m_currentBlock = continuation;
+
+        if (returnType != B3::Void) {
+            Value* phi = continuation->appendNew<Value>(m_proc, Phi, returnType, origin());
+            wasmCallResultUpsilon->setPhi(phi);
+            embedderCallResultUpsilon->setPhi(phi);
+            fillResults(phi);
+        }
+
+    } else {
+
+        Value* patch = createCallPatchpoint(m_currentBlock, origin(), signature, args,
+            scopedLambdaRef<void(PatchpointValue*, Box<PatchpointExceptionHandle>)>([=, this](PatchpointValue* patchpoint, Box<PatchpointExceptionHandle> handle) -> void {
+                patchpoint->effects.writesPinned = true;
+                patchpoint->effects.readsPinned = true;
+
+                // We need to clobber the size register since the LLInt always bounds checks
+                if (m_mode == MemoryMode::Signaling || m_info.memory.isShared())
+                    patchpoint->clobberLate(RegisterSet { PinnedRegisterInfo::get().boundsCheckingSizeRegister });
+                patchpoint->setGenerator([this, handle, unlinkedWasmToWasmCalls, functionIndex, wasmCallerInfo, wasmCalleeInfo](CCallHelpers& jit, const B3::StackmapGenerationParams& params) {
+                    AllowMacroScratchRegisterUsage allowScratch(jit);
+                    prepareForReturnCall(jit, wasmCallerInfo, wasmCalleeInfo);
+                    handle->generate(jit, params, this);
+                    CCallHelpers::Call call = jit.threadSafePatchableNearTailCall();
+                    jit.addLinkTask([unlinkedWasmToWasmCalls, call, functionIndex](LinkBuffer& linkBuffer) {
+                        unlinkedWasmToWasmCalls->append({ linkBuffer.locationOfNearCall<WasmEntryPtrTag>(call), functionIndex });
+                    });
+                });
+            }));
+        fillResults(patch);
+    }
+
+    return { };
+}
+
+void B3IRGenerator::prepareForReturnCall(CCallHelpers& jit, const CallInformation& callerInfo, const CallInformation& calleeInfo)
```

Comment:
same comment I had in the Air version of this function. I also wonder if we can share the code between the two tiers.